kernel/src/fs/epoll/entry.rs

```diff
@@ -0,0 +1,381 @@
+        let is_still_ready = ep_event.is_some()
+            && !inner
+                .flags
+                .intersects(EpollFlags::EDGE_TRIGGER | EpollFlags::ONE_SHOT);
+
+        // If there are events and the epoll entry is one-shot, we need to disable the entry until
+        // the user enables it again via `EpollCtl::Mod`.
+        if ep_event.is_some() && inner.flags.contains(EpollFlags::ONE_SHOT) {
+            self.observer.reset_enabled(&inner);
+        }
+
+        Some((ep_event, is_still_ready))
+    }
+
+    /// Updates the epoll entry by the given event masks and flags.
+    ///
+    /// This method needs to be called in response to `EpollCtl::Add` and `EpollCtl::Mod`.
+    pub(super) fn update(&self, event: EpollEvent, flags: EpollFlags) -> IoEvents {
+        let file = self.file().unwrap();
+
+        let mut inner = self.inner.lock();
+
+        inner.event = event;
+        inner.flags = flags;
+
+        self.observer.set_enabled(&inner);
+
+        file.poll(event.events, Some(&mut inner.poller))
+    }
+
+    /// Shuts down the epoll entry.
+    ///
+    /// This method needs to be called in response to `EpollCtl::Del`.
+    pub(super) fn shutdown(&self) {
+        let mut inner = self.inner.lock();
+
+        self.observer.reset_enabled(&inner);
+        inner.poller.reset();
+    }
+
+    /// Gets the underlying observer.
+    pub(super) fn observer(&self) -> &Observer {
+        &self.observer
+    }
+
+    /// Gets the key associated with the epoll entry.
+    pub(super) fn key(&self) -> &EntryKey {
+        &self.key
+    }
+
+    /// Gets the file descriptor associated with the epoll entry.
+    pub(super) fn fd(&self) -> FileDesc {
+        self.key.fd
+    }
+
+    /// Gets the file associated with this epoll entry.
+    pub(super) fn file_weak(&self) -> &KeyableWeak<dyn FileLike> {
+        &self.key.file
+    }
+}
+
+/// A observer for [`Entry`] that can receive events.
+pub(super) struct Observer {
+    // Whether the entry is enabled.
+    is_enabled: AtomicBool,
+    // Whether the entry is in the ready list.
+    is_ready: AtomicBool,
+    // The ready set of the epoll file that contains this epoll entry.
+    ready_set: Arc<ReadySet>,
+    // The epoll entry itself (always inside an `Arc`).
+    weak_entry: Weak<Entry>,
+}
+
+impl Observer {
+    fn new(ready_set: Arc<ReadySet>, weak_entry: Weak<Entry>) -> Self {
+        Self {
+            is_enabled: AtomicBool::new(false),
+            is_ready: AtomicBool::new(false),
+            ready_set,
+            weak_entry,
+        }
+    }
+
+    /// Returns whether the epoll entry is in the ready list.
+    ///
+    /// *Caution:* If this method is called without holding the lock of the ready list, the user
+    /// must ensure that the behavior is desired with respect to the way the ready list might be
+    /// modified concurrently.
+    pub(super) fn is_ready(&self) -> bool {
+        self.is_ready.load(Ordering::Relaxed)
+    }
```

Comment:
This is because this method may help write fast paths without holding the lock.

But on second thought, I think fast paths may require a memory order stronger than `Relaxed`. Since we don't have a fast path yet, let's require the caller to hold the lock.

Reply:
Requiring the caller to hold the lock looks fine to me.